File: dailyQuestion/2020/2020-07/07-04/python/solution_count.py
def longestValidParentheses(s: str) -> int:
    result = 0
    if not s:
        return result
    n = len(s)
    left, right = 0, 0
    for i in range(n):
        if s[i] == "(":
            left += 1
        else:
            right += 1
        if left == right:
            result = max(result, left + right)
        elif left < right:
            left, right = 0, 0
    left, right = 0, 0
    for i in range(n - 1, -1, -1):
        if s[i] == "(":
            left += 1
        else:
            right += 1
        if left == right:
            result = max(result, left + right)
        if left > right:
            left, right = 0, 0
    return result


if __name__ == "__main__":
    s = "(())()"
    result = longestValidParentheses(s)
    print(result)

File: dailyQuestion/2020/2020-07/07-04/golang/solution_count.go
func longestValidParentheses(s string) int {
	n := len(s)
	result := 0
	if n == 0 {
		return result
	}
	left, right := 0, 0
	for i := 0; i < n; i++ {
		if s[i] == '(' {
			left++
		} else {
			right++
		}
		if left == right {
			result = max(result, left+right)
		} else if right > left {
			left, right = 0, 0
		}
	}
	left, right = 0, 0
	for i := n - 1; i >= 0; i-- {
		if s[i] == '(' {
			left++
		} else {
			right++
		}
		if left == right {
			result = max(result, left+right)
		} else if left > right {
			left, right = 0, 0
		}
	}
	return result
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}